searchTree/searchTree.c

```diff
@@ -0,0 +1,120 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include "searchTree.h"
+
+typedef struct Node {
+    NodeValue value;
+    Node* leftChild;
+    Node* rightChild;
+} Node;
+
+Node* createTreeNode(NodeValue value, int* errorCode) {
+    Node* newNode = calloc(1, sizeof(Node));
+    if (newNode == NULL) {
+        *errorCode = 1;
+        return NULL;
+    }
+
+    newNode->value.key = value.key;
+    newNode->value.value = malloc(strlen(value.value) + 1);
+    if (newNode->value.value == NULL) {
+        free(newNode);
+        *errorCode = 1;
+        return NULL;
+    }
+    strcpy(newNode->value.value, value.value);
+
+    return newNode;
+}
+
+void insertNode(Node* node, NodeValue value, int* error) {
+    if (node->value.key == value.key) {
+        free(node->value.value);
+        node->value.value = malloc(strlen(value.value) + 1);
+        if (node->value.value == NULL) {
+            *error = 1;
+        }
+        else {
+            strcpy(node->value.value, value.value);
+        }
+        return;
+    }
+    if (node->value.key > value.key && !node->leftChild) {
+        Node* newNode = createTreeNode(value, error);
+        if (*error == 1) return;
+        node->leftChild = newNode;
+    }
+    else if (node->value.key < value.key && !node->rightChild) {
+        Node* newNode = createTreeNode(value, error);
+        if (*error == 1) return;
+        node->rightChild = newNode;
+    }
+    else if (node->value.key > value.key) {
+        insertNode(node->leftChild, value, error);
+    }
+    else if (node->value.key < value.key) {
+        insertNode(node->rightChild, value, error);
+    }
+}
+
+Node* getMinNodeOfRightSubtree(Node* node) {
+    Node* parentOfMin = node->rightChild;
+    Node* minNode = parentOfMin->leftChild ? parentOfMin->leftChild : parentOfMin;
+    while (minNode->leftChild) {
+        parentOfMin = minNode;
+        minNode = minNode->leftChild;
+    }
+    parentOfMin->leftChild = minNode->rightChild ? minNode->rightChild : NULL;
+    return minNode;
+}
+
+void deleteNodeByKey(Node* node, int key) {
+    Node* parent = findParentNodeForRemoval(node, key);
+    Node* nodeToDelete = NULL;
+    if (!parent) return;

```

Comment:
```suggestion
    if (!parent) {
        return;
    }
```
Иначе отлаживать такой код неудобно